framework/transport/ieee1905_transport/ieee1905_transport_messages.cpp

```diff
@@ -19,6 +19,132 @@ namespace messages {
+////////////////////////////// Helper Functions //////////////////////////////
+//////////////////////////////////////////////////////////////////////////////
+
+static std::unique_ptr<Message>
+create_transport_message(Type type, std::initializer_list<messages::Message::Frame> frame)
+{
+    switch (messages::Type(type)) {
+    case messages::Type::CmduRxMessage:
+        return std::unique_ptr<messages::CmduRxMessage>{new messages::CmduRxMessage(frame)};
+    case messages::Type::CmduTxMessage:
+        return std::unique_ptr<messages::CmduTxMessage>{new messages::CmduTxMessage(frame)};
+    case messages::Type::SubscribeMessage:
+        return std::unique_ptr<messages::SubscribeMessage>{new messages::SubscribeMessage(frame)};
+    case messages::Type::CmduTxConfirmationMessage:
+        return std::unique_ptr<messages::CmduTxConfirmationMessage>{
+            new messages::CmduTxConfirmationMessage(frame)};
+    case messages::Type::InterfaceConfigurationQueryMessage:
+        return std::unique_ptr<messages::InterfaceConfigurationQueryMessage>{
+            new messages::InterfaceConfigurationQueryMessage(frame)};
+    case messages::Type::InterfaceConfigurationRequestMessage:
+        return std::unique_ptr<messages::InterfaceConfigurationRequestMessage>{
+            new messages::InterfaceConfigurationRequestMessage(frame)};
+    case messages::Type::InterfaceConfigurationIndicationMessage:
+        return std::unique_ptr<messages::InterfaceConfigurationIndicationMessage>{
+            new messages::InterfaceConfigurationIndicationMessage(frame)};
+    default:
+        LOG(WARNING) << "Received unknown message type: " << int(type);
+        return std::unique_ptr<messages::Message>{new messages::Message(0, frame)};
+    }
+}
+
+std::unique_ptr<Message> read_transport_message(Socket &sd)
+{
+    // Peek into the header to check if the entire message received
+    messages::Message::Header header;
+    auto bytes_ready = sd.getBytesReady();
+    auto read_bytes =
+        sd.readBytes(reinterpret_cast<uint8_t *>(&header), sizeof(header), false, sizeof(header));
+
+    if (read_bytes != sizeof(header)) {
+        LOG(ERROR) << "Error peeking into the message header: " << read_bytes;
+        return nullptr;
+    }
+
+    auto discard_pending_bytes = [&]() {
+        if (!sd.getBytesReady()) {
+            return;
+        }
+
+        // Discard all the "ready" bytes in the socket
+        // Reading an invalid message magic means that somehow the data synchronization
+        // was lost. Since the data is not necessarily aligned to any known size, we have
+        // two options here:
+        // 1. Safe - Discard 1 byte at a time, until finding the magic word
+        // 2. Faster - Discard sizeof(Header) bytes and hope to find a valid header afterwads
+        // 2. Fastest - Discard all the bytes and assume the sender will re-send the message
+        // For now we'll use the "Faster" method.
```

Comment:
I totally agree and originally it was implemented using the "Fastest" approach, but to discard the bytes in bigger chunks we need a large enough buffer. Didn't think it's a good idea to have a large static buffer just for this purpose.
Do you think allocating one dynamically in this case is a good compromise?
Unless there's another way to discard the data...

Reply:
The only good way to resync is to look for the magic. And even that is not perfect, because the magic can accidentally happen inside a valid message as well (we're not doing any escaping of the payload).

All this to cover a situation which doesn't even happen in the first place (since we're using a message-oriented connection and we have a hard limit on message size)... So I wouldn't spend too much effort on it.